Init player before incoming message emitted

ihorserba opened the pull request.

ihorserba commented on sdks/js/src/classes/sdk.js:
```diff
@@ -131,6 +131,9 @@ ZCC.Sdk.init({
       sampleRate: 48000
     }, options);
     library.IncomingMessage.PersistentPlayer = new library.Player(playerOptions);
+    library.IncomingMessage.PersistentPlayer.init().catch((err) => {
```
It's problematic.
The deferred in the sdk `init()` resolves once all scripts have been loaded.
And the player's `init()` resolves after a user action.

Maybe you have a better idea than just logging to the console in case of player init failed?

vocoded replied:
well, what i mean is we have:

```
if (shouldInitDefaultPlayer) {
        Sdk.initDefaultPlayer(options);
}
dfd.resolve();
```

and instead we could have:

```
if (shouldInitDefaultPlayer) {
    Sdk.initDefaultPlayer(options).then(() => { dfd.resolve(); }).catch(() => { dfd.reject(); });
}
```

if you return the promise here rather than catch

ihorserba replied:
In this case sdk init is resolved only after player init is resolved.
And the player init resolves after user's gesture.

So for the user it would look like sdk is not loaded until mouse is clicked.

vocoded replied:
i see, that's difficult. i guess the better approach would be to allow the caller to pass a callback in the options which would be invoked on an error

ihorserba replied:
It seems fine to me with the current implementation.
If there was an error on the default player init - it's being logged.

Once incoming message started after that - it's being emitted on `incoming_voice_fail_start`. 

Furthermore, if the default player init was failed on the sdk init - there is another try [on the starting of an incoming message](https://github.com/zelloptt/zello-channel-api/blob/90f9b37544b1582de69fbf673c9b963726d95b59/sdks/js/src/classes/incomingMessage.js#L154-L161).

vocoded replied:
a log statement is not useful programmatically, though. for the other events you mentioned, would it be clear that the root cause is due to the player init failing?

i don't have really strong feelings about this, we don't need to address it now, but it may come up in the future.

ihorserba replied:
Extended the incomingMessage init error, so it says that the problem is within the player init.

ihorserba commented on sdks/js/src/classes/session.js:
```diff
@@ -309,12 +309,17 @@ session.connect(function(err, result) {
+            Constants.EVENT_INCOMING_VOICE_WILL_START,
+            incomingMessage
+          );
+        }).catch((err) => console.error('Failed to init incoming message:', err));
```
Added a new event for this `incoming_voice_failed`

ihorserba commented:
Added uninit of the player once it was failed.

ihorserba commented:
@vocoded please take a look once more

vocoded commented on sdks/js/src/classes/constants.js:
```diff
@@ -48,6 +48,7 @@ module.exports = {
   EVENT_INCOMING_VOICE_WILL_START: 'incoming_voice_will_start',
   EVENT_INCOMING_VOICE_DID_START: 'incoming_voice_did_start',
   EVENT_INCOMING_VOICE_DID_STOP: 'incoming_voice_did_stop',
+  EVENT_INCOMING_VOICE_FAILED: 'incoming_voice_failed',
```
let's call this `incoming_voice_fail_start` to match the existing style of `session_fail_connect`

ihorserba replied:
Renamed

vocoded commented on sdks/js/src/classes/incomingMessage.js:
```diff
@@ -40,16 +40,18 @@ class IncomingMessage extends Emitter {
+      this.initDecoder();
+      this.initSessionHandlers();
+      this.session.on([Constants.EVENT_INCOMING_VOICE_DATA, this.instanceId], this.incomingVoiceHandler);
+      this.session.on([Constants.EVENT_INCOMING_VOICE_DID_STOP, this.instanceId], this.incomingVoiceDidStopHandler);
```
is there any reason these two session handlers should not be added in the `initSessionHandlers` method?

ihorserba replied:
Moved.
Renamed also `initSessionHandlers` -> `initEventHandlers`

ihorserba commented:
Addressed all the comments,
updated pcm-player with the it's latest `master`,
increased the major version,
rebuilt SDK.

It's ready for another look.

vocoded commented on sdks/js/package.json:
```diff
@@ -1,7 +1,7 @@
   "name": "zello-channels-js-sdk",
   "copyright": "Copyright © 2021 Zello Inc.",
-  "version": "0.8.2",
+  "version": "0.9.0",
```
the changes in session are breaking, so we need to bump to 1.0.0

ihorserba replied:
Changed.
I've created a new branch [version_1_0_0](https://github.com/zelloptt/zello-channel-api/tree/version_1_0_0) to merge all the SDK work there. Once it looks good - I'll merge it to master.

ihorserba commented:
Back merged [version_1_0_0](https://github.com/zelloptt/zello-channel-api/tree/version_1_0_0) to [init_player_before_incoming_message_emitted](https://github.com/zelloptt/zello-channel-api/tree/init_player_before_incoming_message_emitted). 
Rebuilt the SDK in order to resolve the conflicts.
